### PR TITLE
JS: http to file access and file to http body taint flow

### PR DESCRIPTION
**Pull request by denislevin**
js: Http to File and File to Http request body taint flow.
An attempt to find code vulnerable to file overwrite with untrusted content or unauthorized arbitrary file upload (information disclosure).
Work in progress.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll [hunk 1]
```diff
@@ -351,11 +351,73 @@ module NodeJSLib {
 
     override DataFlow::Node getAPathArgument() {
       exists (int i | fsFileParam(methodName, i) |
-        result = getArgument(i)
+        result = getArgument(i))
+    }
+  }
+
+  /**
+   * A call to write corresponds to a pattern where file stream is open first, followed by 'read' or 'write' call
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll]: The class doesn't seem to check for a call to `open` or similar. Is that intentional?

File: javascript/ql/src/semmle/javascript/frameworks/Request.qll [hunk 2]
```diff
@@ -44,5 +44,13 @@ module Request {
     }
 
   }
+  
+  // using 'request' library to make http 'POST' and 'PUT' requests with message body. 
+  private class RequestPostBody extends HTTP::RequestBody {
+      RequestPostBody () {
+          this = DataFlow::moduleImport("request").getAMemberCall("post").getArgument(1) or
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/frameworks/Request.qll]: I'd suggest you use `DataFlow::moduleMember("request", "post").getACall()` instead to also handle ES2015-style destructuring imports like this:

```
import { post } from "request";
post(...)
```

File: javascript/ql/src/semmle/javascript/security/dataflow/FileAccessToHttp.qll [hunk 3]
```diff
@@ -0,0 +1,87 @@
+/** 
+ * Provides Taint tracking configuration for reasoning about file access taint flow to html post body 
+ */
+import javascript
+import semmle.javascript.frameworks.HTTP
+
+module FileAccessToHttpDataFlow {
+  /**
+   * A data flow source for code injection vulnerabilities.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/security/dataflow/FileAccessToHttp.qll]: Here and below: "code injection" seems to be copied from another query.

File: javascript/ql/src/semmle/javascript/security/dataflow/FileAccessToHttp.qll [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    override predicate isSanitizer(DataFlow::Node node) {
+      super.isSanitizer(node) or
+      node instanceof Sanitizer
+    }
+  }
+
+/** Taint step propagating a tainted parameter or a  variable to a wrapper object. */ 
+  private class ObjectTaintStep extends TaintTracking::AdditionalTaintStep {
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/security/dataflow/FileAccessToHttp.qll]: Adding this taint step can lead to significant imprecision: if `v` is tainted and you have an assignment `o.p = v`, then `o` will become tainted, so any (!) read of a property (not just `p`) from `o` will also be considered tainted. Do you have a concrete use case in mind that you want to model?

File: javascript/ql/src/semmle/javascript/security/dataflow/FileAccessToHttp.qll [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
+      pred = this and succ = target
+    }
+  }  
+
+  /** A source is a file access parameter, as in readFromFile(buffer). */
+  private class FileAccessArgumentAsSource extends Source {
+    FileAccessArgumentAsSource() {  
+      exists(DataFlow::CallNode cn |
+        cn instanceof FileSystemAccess and
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/security/dataflow/FileAccessToHttp.qll]: This essentially breaks the abstraction provided by `FileSystemAccess` by casting it to `CallNode`. I'd suggest adding the required API on `FileSystemAccess` instead.

File: javascript/ql/src/semmle/javascript/security/dataflow/FileAccessToHttp.qll [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        this = cn.getAnArgument()
+      )
+    }
+  }
+
+  /** A source is a file access callback parameter, as in fs.readFile('MyFile.json', 'utf-8', function (err, data) {}. */
+  private class FileAccessParamAsSource extends Source {
+    FileAccessParamAsSource() {  
+      this instanceof FileSystemAccess
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/security/dataflow/FileAccessToHttp.qll]: This seems overly broad. Perhaps you'd want to restrict this to certain kinds of file system accesses?

File: javascript/ql/src/semmle/javascript/security/dataflow/FileAccessToHttp.qll (hunk 4 above)
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/security/dataflow/FileAccessToHttp.qll]: I'll also weigh in here that `AdditionalTaintStep` should not be subclassed in a query-specific file; there is a performance penalty when doing so. That class is only for taint steps that are shared by all taint configurations.

Steps that are specific to a configuration should be added in `isAdditionalTaintStep` (even if it's the only configuration used by the query).

File: javascript/ql/src/semmle/javascript/security/dataflow/FileAccessToHttp.qll (hunk 5 above)
Comment (denislevin) [reply javascript/ql/src/semmle/javascript/security/dataflow/FileAccessToHttp.qll]: This source is specifically dealing with FileSystemAccess that is also a CallNode. For these, the source is an argument.
The one below is FileSystemAccess that is a Parameter (as in a callback), or anything else.

File: javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll (hunk 1 above)
Comment (denislevin) [reply javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll]: I've split this class in two and now I can check for createReadStream and createWriteStream methods

**Comment by denislevin** [comment]
Guys, need to ask for help here. I don't know why the build is failing, and link with "Details" doesn't open for me (https://jenkins.internal.semmle.com/job/Language-Tests/job/JavaScript/3697/ not leading anywhere).
I *suspect* this is because my Windows git changed line endings on some of the tests to \r\n instead of \n. 
'ql\.editorconfig has this:
[*.{ql,qll,qlref,dbscheme,qhelp}]
end_of_line = lf
`
so test files (*.cpp, *.cs, *.js, *.ts, etc. ) are not included.

**Comment by asger-semmle** [comment]
Sorry about the hassle. It's a compilation error due to the PR going out of sync with master:
```
ERROR: Could not resolve type HttpRequest (ql/javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll:665,14-25)
```
The internal class `HttpRequest` was subsumed by `ClientRequest`. The old name didn't make it clear if this was a request being sent or a request being received, which is an important distinction. We do try to be careful about breaking changes, but this was an internal class.

Could you try rebasing or merging in master? Renaming `HttpRequest` to `NodeJSClientRequest` should then do the trick.

**Comment by asger-semmle** [comment]
Could you please also remove the "Normalize line endings" commit? It touches files that shouldn't be modified here, and it seems to insert CR line endings which we are trying to avoid (https://github.com/Semmle/ql/pull/181).

**Comment by denislevin** [comment]
Thank you for the review and error root cause! I have fixed it and reverted "Normalize line endings" commit.

**Comment by denislevin** [comment]
Got my change refactored to take care of separating read from write file access where applicable.

**Comment by xiemaisi** [comment]
Thanks for your latest updates, @denislevin! If you think this is more or less ready, we'd be happy to merge and perhaps fix up some minor things later on.

For the third-party code snippets used in your tests, could you add a header comment like [this one](https://github.com/Semmle/ql/blob/master/javascript/ql/test/library-tests/frameworks/SQL/mssql1.js#L1) indicating where you got them from?

**Comment by denislevin** [comment]
Done! Yes, I think it's ready to merge.

**Comment by denislevin** [comment]
Please merge if you see it fit.
Thank you!
=Denis

**Comment by xiemaisi** [comment]
Thank you for your contribution, @denislevin!

**Pull request merged**